### DISCUSSION TITLE
How to access grid coordinates when defining the evolution_rate of a customed PDE class?

### DISCUSSION THREAD

**Discussion by mynanshan** [zwicker-group/py-pde#566] [Q&A] [answered]
This might be a dumb question, but what is the internal approach to get the values of the grid?  

Specifically, when writing the member function `evolution_rate` of a customer PDE class, I need to use the values of the grid. Although I can just calculate the grid values in advance outside the PDE class, I think there should be a way to access the grid values from the argument `state` or anything internal.

**Comment by david-zwicker** [comment]
I'm a not entirely sure what you mean by "values of the grid", but you can access the grid object using `state.grid` and if you need direct access to the field values at the grid cells, you can access the numpy array `state.data`. Does this help?

**Reply by mynanshan** [reply]
Sorry that I have not made it clear. I was meant to say the coordinates of the grid. Suppose the pde is c(x). Then, if I got it right, state.data should be the value of c evaluated on the grid, but I want to use the discrete values of x.

**Reply by david-zwicker** [reply, marked as answer]
You can use `state.grid.axes_coords`, `state.grid.cell_coords`, or `state.grid.coordinate_arrays`; see [the documentation](https://py-pde.readthedocs.io/en/latest/packages/pde.grids.base.html#pde.grids.base.GridBase) for the exact details.
